### PR TITLE
feature(rpc): add real data to `getinfo` method

### LINKED ISSUE #3142: getinfo JSON-RPC method
## Motivation

`lightwalletd` uses the `getinfo` JSON-RPC method.

### Required Fields

The method is documented here: https://zcash.github.io/rpc/getinfo.html

But only the following fields are used by `lightwalletd`: https://github.com/zcash/lightwalletd/blob/master/common/common.go#L91

Field list:
```
{
  "build": xxxxx,           (string) `zcashd`'s CLIENT_BUILD
  "subversion": xxxxx,   (string) Subversion as sent to the P2P network in `version` messages - `zcashd`'s CLIENT_NAME, CLIENT_VERSION, uacomments
}
```

### Tasks

- Implement the RPC method
    - build: `app_version()`
    - subversion: `USER_AGENT`
- Add RPC acceptance tests to CI
- Test that the RPC method works with `lightwalletd` 

## API Reference

We plan to use [jsonrpc_core](https://docs.rs/jsonrpc-core/18.0.0/jsonrpc_core/) with:
* [jsonrpc_http_server](https://github.com/paritytech/jsonrpc/tree/master/http)
* [jsonrpc-derive](https://github.com/paritytech/jsonrpc/tree/master/derive)

### Example Code

Here are examples of:
* [declaring RPC methods as Rust functions](https://github.com/paritytech/jsonrpc#basic-usage-with-derive)
* [testing RPC methods](https://github.com/paritytech/jsonrpc/blob/master/derive/examples/client-local.rs)

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

Implement the `getinfo` RPC method with the fields used by lightwalletd. Close https://github.com/ZcashFoundation/zebra/issues/3142 if merged.

## Solution

I had a hard time trying to pass the app version into the method. This approach is with `lazy_static`. I am open to change this if i find something that works and it is better.

This PR does not have a specific test, i am unsure if the tests we will use will be acceptance or something else for each method. If they are acceptance then the test for this call should be added after https://github.com/ZcashFoundation/zebra/pull/3641 as it is very similar to what we have now.

The code in this pull was tested manually:

```
$ curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc": "2.0", "method": "getinfo", "params": [], "id":123 }' 127.0.0.1:8832 | jq
{
  "jsonrpc": "2.0",
  "result": {
    "build": "1.0.0-beta.5+22.g45caa63",
    "subversion": "/Zebra:1.0.0-beta.5/"
  },
  "id": 123
}
$ 
```

The docs for the method are in a discussion at https://github.com/ZcashFoundation/zebra/discussions/3638. Open to change the format.

## Review

Probably @teor2345 will be the best option if they have time so i can have some feedback for my morning.

### Reviewer Checklist

  - [ ] Lazy static implementation is acceptable.
  - [ ] If so, check the code as a whole.

### CODE REVIEW

**Comment by teor2345** [comment]
> This PR does not have a specific test, i am unsure if the tests we will use will be acceptance or something else for each method.

@oxarbitrage this seems like it might be a decision for the whole team. Did you want to put it on the engineering sync agenda for tomorrow?

File: zebra-rpc/src/methods.rs [hunk 1]
```diff
@@ -9,15 +9,31 @@
...
 pub trait Rpc {
     /// getinfo
     ///
-    /// TODO: explain what the method does
-    ///       link to the zcashd RPC reference
-    ///       list the arguments and fields that lightwalletd uses
-    ///       note any other lightwalletd changes
+    /// Returns software information from the RPC server running Zebra.
+    ///
+    /// zcadhd reference: <https://zcash.github.io/rpc/getinfo.html>
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs]:
```suggestion
    /// zcashd reference: <https://zcash.github.io/rpc/getinfo.html>
```

File: zebra-rpc/src/methods.rs [hunk 2]
```diff
@@ -9,15 +9,31 @@
...
+    /// Result:
+    /// {
+    ///      "build": String, // Full application version
+    ///      "subversion", String, // Zebra user agent
+    /// }
+    ///
+    /// Note 1: We only expose 2 fields as they are the only ones needed for
+    /// lightwalletd: <https://github.com/zcash/lightwalletd/blob/v0.4.9/common/common.go#L91-L95>
+    ///
+    /// Note 2: <https://zcash.github.io/rpc/getinfo.html> is outdated so it does not
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs:35]: Optional: Can you open a ticket at https://github.com/zcash/zcash/issues for this?

The zcashd team have asked us to open tickets for any missing RPC documentation.
(There's no need to do a PR, just let them know which fields are missing.)

File: zebra-rpc/src/methods.rs [hunk 3]
```diff
@@ -35,10 +51,15 @@ pub trait Rpc {
 pub struct RpcImpl;
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Suggestion (1/3): storing the string in the RPC method object seems a bit simpler than a lazy_static and mutex.

```suggestion
pub struct RpcImpl {
    /// Zebra's application version.
    pub app_version: String,
}
```

File: zebra-rpc/src/methods.rs [hunk 4]
```diff
@@ -35,10 +51,15 @@ pub trait Rpc {
...
 impl Rpc for RpcImpl {
     fn get_info(&self) -> Result<GetInfo> {
-        // TODO: dummy output data, fix in the context of #3142
         let response = GetInfo {
-            build: "TODO: Zebra v1.0.0 ...".into(),
-            subversion: "TODO: /Zebra:1.0.0-beta.../".into(),
+            build: String::from_utf8(
```
Comment (teor2345) [inline zebra-rpc/src/methods.rs]: Suggestion (3/3): return a copy of the string from the RPC method object

```suggestion
            build: self.app_version.clone(),
```

File: zebra-rpc/src/server.rs [hunk 5]
```diff
@@ -22,9 +23,14 @@ pub mod compatibility;
...
+    /// Storage for the application version string coming from zebrad.
+    pub static ref APP_VERSION: Mutex<Vec<u8>> = Mutex::new(vec![]);
+}
+
 impl RpcServer {
     /// Start a new RPC server endpoint
-    pub fn spawn(config: Config) -> tokio::task::JoinHandle<()> {
+    pub fn spawn(config: Config, app_version: String) -> tokio::task::JoinHandle<()> {
         if let Some(listen_addr) = config.listen_addr {
             info!("Trying to open RPC endpoint at {}...", listen_addr,);
```
Comment (teor2345) [inline zebra-rpc/src/server.rs:30]: Suggestion (2/3): initialise the the RPC method object with the string.

You'll also need to pass `rpc_impl` to `extend_with`. (GitHub won't let me do suggestions on that line.)

```suggestion

let rpc_impl = RpcImpl { app_version };

```

**Review by teor2345 — COMMENTED** [review]
I haven't tried this myself, but I think it might simplify things.

File: zebra-rpc/src/methods.rs (hunk 1 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs]: Fixed this and added a ticket to zcashd here https://github.com/zcash/zcash/issues/5606

File: zebra-rpc/src/methods.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods.rs:35]: https://github.com/ZcashFoundation/zebra/pull/3660/commits/dd0b18fa18c20e220c9981b12f5ebc65cafd7859

File: zebra-rpc/src/server.rs (hunk 5 above)
Comment (oxarbitrage) [reply zebra-rpc/src/server.rs:30]: Thanks, this is the part i had missing when i did it this way. I refactored everything at https://github.com/ZcashFoundation/zebra/pull/3660/commits/dd0b18fa18c20e220c9981b12f5ebc65cafd7859

**Review by teor2345 — COMMENTED** [review]
Looks good, let's check what tests we need with the team before we merge.

**Comment by oxarbitrage** [comment]
Lets add some basic test here before we merge. This will also define the test structure for all the methods.

**Review by teor2345 — APPROVED** [review]
Looks great, thanks!

**Pull request merged**